src/blockchain/log-processors/order-filled.ts

```diff
@@ -1,17 +1,167 @@
+  outcome: number;
+}
+
+interface MarketsRow {
+  minPrice: string|number;
+  maxPrice: string|number;
+  numTicks: string|number;
+}
+
+interface PositionsRow {
+  outcome: number;
+  marketID?: Address;
+  numShares?: string|number;
+  account?: Address;
+  realizedProfitLoss?: string|number;
+  unrealizedProfitLoss?: string|number;
+  numSharesAdjustedForUserIntention?: string|number;
+}
+
+interface OutcomesRow {
   price: string|number;
-  maker: boolean; // NB: UI expects "maker" field to be true if the current user is the creator of the order, false otherwise
-  tradeGroupID: Bytes32|null;
+}
+
+export function calculateNumberOfSharesTraded(numShares: string, numTokens: string, price: string): string {
+  return new BigNumber(numShares, 10).plus(new BigNumber(numTokens, 10).dividedBy(new BigNumber(price, 10))).toFixed();
+}
+
+export function insertPositionInMarket(db: Knex, trx: Knex.Transaction, account: Address, marketID: Address, positionInMarket: Array<string>, realizedProfitLoss: Array<string>, unrealizedProfitLoss: Array<string>, positionInMarketAdjustedForUserIntention: Array<string>, callback: ErrorCallback): void {
+  db.batchInsert("positions", positionInMarket.map((numShares: string, outcome: number): PositionsRow => ({
+    account,
+    marketID,
+    outcome,
+    numShares,
+    realizedProfitLoss: realizedProfitLoss[outcome],
+    unrealizedProfitLoss: unrealizedProfitLoss[outcome],
+    numSharesAdjustedForUserIntention: positionInMarketAdjustedForUserIntention[outcome],
+  })), 1000).transacting(trx).asCallback(callback);
+}
+
+export function updatePositionInMarket(db: Knex, trx: Knex.Transaction, account: Address, marketID: Address, positionInMarket: Array<string>, realizedProfitLoss: Array<string>, unrealizedProfitLoss: Array<string>, positionInMarketAdjustedForUserIntention: Array<string>, callback: ErrorCallback): void {
+  forEachOf(positionInMarket, (numShares: string, outcome: number, nextOutcome: AsyncCallback): void => {
+    db("positions").transacting(trx).where({ account, marketID, outcome }).update({
+      numShares,
+      realizedProfitLoss: realizedProfitLoss[outcome],
+      unrealizedProfitLoss: unrealizedProfitLoss[outcome],
+      numSharesAdjustedForUserIntention: positionInMarketAdjustedForUserIntention[outcome],
+    }).asCallback(nextOutcome);
+  }, callback);
+}
+
+export function calculateProfitLossInOutcome(augur: Augur, trx: Knex.Transaction, account: Address, marketID: Address, outcome: number, callback: (err: Error|null, profitLossInOutcome?: any) => void): void {
+  getUserTradingHistory(trx, account, marketID, outcome, null, null, null, null, null, (err: Error|null, userTradingHistory?: Array<UITrade>): void => {
+    if (err) return callback(err);
+    trx.select("price").from("outcomes").where({ marketID, outcome }).asCallback((err: Error|null, outcomesRows?: Array<OutcomesRow>): void => {
+      if (err) return callback(err);
+      const { price } = outcomesRows![0];
```

Comment:
Should there be a check here for `!outcomesRows || !outcomesRows.length` ?

Reply:
Hm, I'm not 100% clear what all the combinations of things that can be returned from the knex methods are -- but yes that'd be good defensive practice.  I'll add!

Reply:
Yea, i'd love to take some time and play around with Knex to see what values it callback's with in various circumstances! It might never be undefined/null here, and we could clean up a lot of code.